payroll/models/hr_payslip.py

```diff
@@ -0,0 +1,581 @@
+                # delete old worked days lines
+                "worked_days_line_ids": [(2, x) for x in self.worked_days_line_ids.ids],
+                # 'details_by_salary_head':[], TODO put me back
+                "name": "",
+                "contract_id": False,
+                "struct_id": False,
+            }
+        }
+        if (not employee_id) or (not date_from) or (not date_to):
+            return res
+        ttyme = datetime.combine(fields.Date.from_string(date_from), time.min)
+        employee = self.env["hr.employee"].browse(employee_id)
+        locale = self.env.context.get("lang") or "en_US"
+        res["value"].update(
+            {
+                "name": _("Salary Slip of %s for %s")
+                % (
+                    employee.name,
+                    tools.ustr(
+                        babel.dates.format_date(
+                            date=ttyme, format="MMMM-y", locale=locale
+                        )
+                    ),
+                ),
+                "company_id": employee.company_id.id,
+            }
+        )
+
+        if not self.env.context.get("contract"):
+            # fill with the first contract of the employee
+            contract_ids = self.get_contract(employee, date_from, date_to)
+        else:
+            if contract_id:
+                # set the list of contract for which the input have to be
+                # filled
+                contract_ids = [contract_id]
+            else:
+                # if we don't give the contract, then the input to fill should
+                # be for all current contracts of the employee
+                contract_ids = self.get_contract(employee, date_from, date_to)
+
+        if not contract_ids:
+            return res
+        contract = self.env["hr.contract"].browse(contract_ids[0])
+        res["value"].update({"contract_id": contract.id})
+        struct = contract.struct_id
+        if not struct:
+            return res
+        res["value"].update({"struct_id": struct.id})
+        # computation of the salary input
+        contracts = self.env["hr.contract"].browse(contract_ids)
+        worked_days_line_ids = self.get_worked_day_lines(contracts, date_from, date_to)
+        input_line_ids = self.get_inputs(contracts, date_from, date_to)
+        res["value"].update(
+            {
+                "worked_days_line_ids": worked_days_line_ids,
+                "input_line_ids": input_line_ids,
+            }
+        )
+        return res
+
+    @api.onchange("employee_id", "date_from", "date_to")
+    def onchange_employee(self):
+
+        if (not self.employee_id) or (not self.date_from) or (not self.date_to):
+            return
+
+        employee = self.employee_id
+        date_from = self.date_from
+        date_to = self.date_to
+        contract_ids = []
+
+        ttyme = datetime.combine(fields.Date.from_string(date_from), time.min)
```

Comment:
```suggestion
        ttyme = datetime.combine(date_from, time.min)
```
version 12+ Odoo returns always native Python objects

ref : https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-12.0#tasks-to-do-in-the-migration